app/upgrades/v16/upgrades.go

```diff
@@ -72,3 +81,38 @@ func CreateUpgradeHandler(
 		return mm.RunMigrations(ctx, configurator, vm)
 	}
 }
+
+// ConvertToNativeCoinExtensions converts all the registered ERC20 tokens of Cosmos native token pairs
+// back to the native representation and registers the (W)ERC20 precompiles for each token pair.
+func ConvertToNativeCoinExtensions(
+	ctx sdk.Context,
+	logger log.Logger,
+	accountKeeper authkeeper.AccountKeeper,
+	bankKeeper bankkeeper.Keeper,
+	erc20Keeper erc20keeper.Keeper,
+) error {
+	// NOTE (@fedekunze): first we must convert the all the registered tokens.
+	// If we do it the other way around, the conversion will fail since there won't
+	// be any contract code due to the selfdestruct.
+	if err := ConvertERC20Coins(ctx, logger, accountKeeper, bankKeeper, erc20Keeper); err != nil {
+		logger.Error("failed to convert native coins", "error", err.Error())
+		// TODO: return error here?
+		// return errorsmod.Wrap(err, "failed to convert native coins")
+	}
+
+	// Instantiate the (W)ERC20 Precompile for each registered IBC Coin
+
+	// IMPORTANT (@fedekunze): This logic needs to be included on EVERY UPGRADE
+	// from now on because the AvailablePrecompiles function does not have access
+	// to the state (in this case, the registered token pairs).
+	//
+	// FIXME: Do we want to run both if the convert coins failed? I think that could be dangerous if a coin
+	// was not fully converted and we overwrite the contract address with the Extension? Or maybe not?
```

Comment:
can you elaborate @MalteHerrmann

Reply:
if I understand it correctly, we are deleting the contract storage for all of the ERC20 contracts:

https://github.com/evmos/evmos/blob/ca0f3e4d3e8407fd983b42aa595b052a67c1598b/x/erc20/keeper/precompiles.go#L47-L55

This would mean, that if the conversion was not successful for some ERC20 representations those tokens would only be present in the contract storage right? And if we delete the contract code and just replace it with the EVM extension that holds no internal storage, all of the non-converted balances would be wiped right?

Reply:
it might even make sense to pass a cached context to the `ConvertToNativeCoinExtensions` function and only execute the `writeFn` if it's successful 🤔

Reply:
cc @fedekunze

Reply:
Okay I wrapped the execution with a cached context, so in case of any error we don't write the changes to state. We'll have to check this with the mainnet data and check if any errors would occur. In tests with the integration test utils I couldn't produce a state that would raise any errors.